pdns/ws-auth.cc

```diff
@@ -1079,81 +1079,88 @@ static void apiServerSearchData(HttpRequest* req, HttpResponse* resp) {
     throw HttpMethodNotAllowedException();
 
   string q = req->getvars["q"];
+  string sMax = req->getvars["max"];
+  int maxEnts = 100;
+  int ents = 0;
+
   if (q.empty())
     throw ApiException("Query q can't be blank");
+  if (sMax.empty() == false)
+    maxEnts = boost::lexical_cast<int>(sMax);
+  if (maxEnts < 1)
+    throw ApiException("Maximum entries must be larger than 0");
 
+  SimpleMatch sm(q,true);
   UeberBackend B;
-
   vector<DomainInfo> domains;
-  B.getAllDomains(&domains, true); // incl. disabled
-
+  vector<DNSResourceRecord> result_rr;
+  vector<Comment> result_c;
+  map<int,DNSName> zoneIdZone;
+  map<int,DNSName>::iterator val;
   Document doc;
-  doc.SetArray();
 
-  DNSResourceRecord rr;
-  Comment comment;
+  doc.SetArray();
 
-  BOOST_FOREACH(const DomainInfo& di, domains) {
-    string zoneId = apiZoneNameToId(di.zone.toString());
+  B.getAllDomains(&domains, true);
 
-    if (pdns_ci_find(di.zone.toString(), q) != string::npos) {
+  for(const DomainInfo di: domains)
+  {   
+    if (ents < maxEnts && sm.match(di.zone)) {
       Value object;
       object.SetObject();
-      object.AddMember("type", "zone", doc.GetAllocator());
-      Value jzoneId(zoneId.c_str(), doc.GetAllocator()); // copy
-      object.AddMember("zone_id", jzoneId, doc.GetAllocator());
+      object.AddMember("object_type", "zone", doc.GetAllocator());
+      object.AddMember("zone_id", di.id, doc.GetAllocator());
       Value jzoneName(di.zone.toString().c_str(), doc.GetAllocator()); // copy
       object.AddMember("name", jzoneName, doc.GetAllocator());
       doc.PushBack(object, doc.GetAllocator());
+      ents++;
     }
+    zoneIdZone[di.id] = di.zone; // populate cache
+  }
 
-    // if zone name is an exact match, don't bother with returning all records/comments in it
-    if (di.zone == q) {
-      continue;
-    }
-    // the code below is too slow
-#if 0
-    di.backend->list(di.zone, di.id, true); // incl. disabled
-    while(di.backend->get(rr)) {
-      if (!rr.qtype.getCode())
-        continue; // skip empty non-terminals
-
-      if (pdns_ci_find(rr.qname, q) == string::npos && pdns_ci_find(rr.content, q) == string::npos)
-        continue;
-
+  if (B.searchRecords(q, maxEnts, result_rr))
```

Comment:
false indicates that the search failed for some reason, not that it did not give results. result_rr will be unspecified in that case.